README

General Information
===================

This directory contains libufodecode version 0.6. This package provides a
library to decode frames from the UFO camera developed at IPE/KIT. It is able to
decode frames produced with firmware versions 4, 5 and 5 with 12-bit support.
The library is a dependency of pcilib to decode frames on-the-fly.

This package also contains a stand-alone offline decoder called `ipedec` to
decode raw frames acquired with the `pcitool` program. More information is
available by calling

  $ ipedec -h

Installation
============

Please see the file called INSTALL.


Licensing
=========

Please see the file called COPYING.


NEWS
Changes in ufodecode 0.2.3
==========================

This is bugfix release:

- Fix segmentation fault for large sensor
- Try to guess the number of rows from header
- Fix installation


Changes in ufodecode 0.2.2
==========================

This is a bugfix release:

- Fix 10 bit decoding that shows wrong output when ADCs are almost saturated
- Fix 12 bit decoding that capped some output channels at 10 bits
- `ipedec` returns error code different from 0 when an error occurred
- `ipedec` reports the frame rate again with the `--frame-rate` switch set


Changes in ufodecode 0.2.1
==========================

This is a bugfix release:

- Fix 12 bit decoding according to CMOSIS specs
- Fix command line input of `ipedec`
- Truncate so-Version to major.minor


Changes in ufodecode 0.2.0
==========================

Name change
-----------
In accordance to the rest of the software, the type names changed from
`ufo_decoder_t` to `UfoDecoder`. Moreover there is no typedef for pointers to
`UfoDecoder` structures.


API breakage
------------
To simplify debugging, all status data is passed from the decoded frame to the
caller of `ufo_decoder_decode_frame()` and `ufo_decoder_get_next_frame()` by
means of the new `UfoDecoderMeta` structure as defined in `ufodecode.h`.


Minor changes
-------------
- The decoder supports version 4 and 5 of the UFO camera frame format.
- `ipedec` now has a proper command line interface.
- `ipedec` prints the frame meta data when using the verbose command line switch
  (`-v` or `--verbose`).


test/timer.c
#include <stdlib.h>
#include "timer.h"

struct _Timer {
    struct timeval  start;
    long            seconds;
    long            useconds;
};


Timer *
timer_new (void)
{
    Timer *t = (Timer *) malloc (sizeof (Timer));
    t->seconds = t->useconds = 0L;
    return t;
}

void
timer_destroy (Timer *t)
{
    free (t);
}

void
timer_start (Timer *t)
{
    gettimeofday(&t->start, NULL);
}

void
timer_stop (Timer *t)
{
    struct timeval end;

    gettimeofday(&end, NULL);
    t->seconds += end.tv_sec - t->start.tv_sec;
    t->useconds += end.tv_usec - t->start.tv_usec;
}

double
timer_get_seconds (Timer *t)
{
    return t->seconds + t->useconds / 1000.0 / 1000.0;
}


test/timer.h
#ifndef TIMER_H
#define TIMER_H

#include <sys/time.h>

typedef struct _Timer Timer;

Timer * timer_new           (void);
void    timer_destroy       (Timer *t);
void    timer_start         (Timer *t);
void    timer_stop          (Timer *t);
double  timer_get_seconds   (Timer *t);

#endif


test/ipedec.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <getopt.h>
#include <ufodecode.h>
#include "timer.h"

static const int MAX_ROWS = 2048;

typedef struct {
    int clear_frame;
    int dry_run;
    int verbose;
    int rows;
    int print_frame_rate;
    int print_num_rows;
    int cont;
    int convert_bayer;
} Options;


static int
read_raw_file(const char *filename, char **buffer, size_t *length)
{
    FILE *fp = fopen(filename, "rb");
    if (fp == NULL)
        return ENOENT;

    fseek(fp, 0, SEEK_END);
    *length = ftell(fp);
    rewind(fp);

    *buffer = (char *) malloc(*length);

    if (*buffer == NULL) {
        fclose(fp);
        return ENOMEM;
    }

    size_t buffer_length = fread(*buffer, 1, *length, fp);
    fclose(fp);
    if (buffer_length != *length) {
        free(*buffer);
        return ERANGE;
    }
    return 0;
}

static void
usage(void)
{
    printf("usage: ipedec [OPTION]... FILE [FILE ...]\n\
Options:\n\
  -h, --help                Show this help message and exit\n\
  -v, --verbose             Print additional information on STDOUT\n\
  -r, --num-rows=N          N rows that are contained in the file\n\
  -c, --clear-frame         Clear the frame for each iteration\n\
  -d, --dry-run             Do not save the frames\n\
  -f, --print-frame-rate    Print frame rate on STDOUT\n\
      --print-num-rows      Print number of rows on STDOUT\n\
      --continue            Continue decoding frames even when errors occur\n\
      --convert-bayer       Convert Bayer pattern to 24 Bit RGB\n");
}

static void
print_meta_data (UfoDecoderMeta *meta)
{
    printf("  frame_number    = %i\n", meta->frame_number);
    printf("  time_stamp      = %i\n", meta->time_stamp);
    printf("  output_mode     = %i\n", meta->output_mode);
    printf("  adc_resolution  = %i\n", meta->adc_resolution);
    printf("  n_rows          = %i\n", meta->n_rows);
    printf("  n_skipped_rows  = %i\n", meta->n_skipped_rows);

    printf("  status1\n");
    printf("    fsm_master_readout = %i\n", meta->status1.desc.fsm_master_readout);
    printf("    fsm_daq         = %i\n", meta->status1.desc.fsm_daq);
    printf("    pixel_full      = %i\n", meta->status1.desc.pixel_full);
    printf("    control_lock    = %i\n", meta->status1.desc.control_lock);
    printf("    data_lock       = %i\n", meta->status1.desc.data_lock);

    printf("  status2\n");
    printf("    end_of_frames   = %i\n", meta->status2.desc.end_of_frames);
    printf("    busy_or         = %i\n", meta->status2.desc.busy_or);
    printf("    busy_ddr        = %i\n", meta->status2.desc.busy_ddr);
    printf("    busy_interl     = %i\n", meta->status2.desc.busy_interl);
    printf("    error_status    = %i\n", meta->status2.desc.error_status);
    printf("    data_fifo_read_count = %i\n", meta->status2.desc.data_fifo_read_count);
    printf("    data_fifo_full       = %i\n", meta->status2.desc.data_fifo_full);
    printf("    data_fifo_empty      = %i\n", meta->status2.desc.data_fifo_empty);
    printf("    ddr_fifo_write_count = %i\n", meta->status2.desc.ddr_fifo_write_count);
    printf("    ddr_fifo_full        = %i\n", meta->status2.desc.ddr_fifo_full);
    printf("    ddr_fifo_empty       = %i\n", meta->status2.desc.ddr_fifo_empty);

    printf("  status3\n");
    printf("    row_counter     = %i\n", meta->status3.desc.row_counter);
    printf("    pixel_counter   = %i\n", meta->status3.desc.pixel_counter);
    printf("    ddr_read        = %i\n", meta->status3.desc.ddr_read);
    printf("    ddr_write       = %i\n", meta->status3.desc.ddr_write);
    printf("    ddr_arbiter     = %i\n", meta->status3.desc.ddr_arbiter);
    printf("\n");
}

static void
write_raw_file (UfoDecoderMeta *meta,
                Options *opts,
                uint16_t *pixels,
                FILE *fp)
{
    size_t n_rows = meta->n_rows;

    if (opts->convert_bayer) {
        uint8_t *rgb_pixels = malloc (2048 * n_rows * 3);

        ufo_convert_bayer_to_rgb (pixels, rgb_pixels, 2048, n_rows);
        fwrite (rgb_pixels, sizeof(uint8_t), 2048 * n_rows * 3, fp);
        free (rgb_pixels);
    }
    else {
        fwrite(pixels, sizeof(uint16_t), 2048 * n_rows, fp);
    }
}

static int
process_file(const char *filename, Options *opts)
{
    UfoDecoder      *decoder;
    UfoDecoderMeta   meta = {0};
    Timer           *timer;
    char            *buffer;
    size_t           num_bytes;
    uint16_t        *pixels;
    uint32_t         time_stamp, old_time_stamp;
    int              n_frames;
    int              error = 0;
    FILE            *fp;
    char             output_name[256];
    float            mtime;

    error = read_raw_file(filename, &buffer, &num_bytes);

    if (error) {
        fprintf(stderr, "Error reading %s: %s\n", filename, strerror(error));
        return error;
    }

    decoder = ufo_decoder_new(opts->rows, 2048, (uint32_t *) buffer, num_bytes);

    if (decoder == NULL) {
        fprintf(stderr, "Failed to initialize decoder\n");
        return 1;
    }

    if (!opts->dry_run) {
        snprintf(output_name, 256, "%s.raw", filename);
        fp = fopen(output_name, "wb");

        if (!fp) {
            fprintf(stderr, "Failed to open file for writing\n");
            return 1;
        }
    }

    timer = timer_new ();
    pixels = (uint16_t *) malloc (2048 * MAX_ROWS * sizeof(uint16_t));
    n_frames = 0;
    old_time_stamp = 0;

    while (error != EIO) {
        timer_start (timer);
        error = ufo_decoder_get_next_frame (decoder, &pixels, &meta);

        if (meta.n_rows == 0)
            meta.n_rows = opts->rows;

        timer_stop (timer);
        n_frames++;

        if (!error) {
            if (opts->verbose) {
                printf("Status for frame %i\n", n_frames);
                print_meta_data (&meta);
            }

            if (opts->print_frame_rate) {
                uint32_t diff = 80 * (meta.time_stamp - old_time_stamp);

                printf("%-6d", 1000000000 / diff);
                old_time_stamp = meta.time_stamp;
            }

            if (opts->print_num_rows)
                printf("%d", meta.n_rows); 

            if (opts->print_frame_rate || opts->print_num_rows)
                printf("\n");

            if (opts->clear_frame)
                memset(pixels, 0, 2048 * meta.n_rows * sizeof(uint16_t));

            if (!opts->dry_run)
                write_raw_file (&meta, opts, pixels, fp);
        }
        else if (error != EIO) {
            fprintf(stderr, "Failed to decode frame %i\n", n_frames);

            if (opts->cont) {
                /* Save the frame even though we know it is corrupted */
                if (!opts->dry_run)
                    write_raw_file (&meta, opts, pixels, fp);
            }
            else
                break;
        }
    }

    if (!opts->dry_run)
        fclose(fp);

    if (opts->verbose) {
        printf("Decoded %i frames in %.5fms\n", n_frames,
               timer_get_seconds (timer) * 1000.0);
    }

    free(pixels);
    free(buffer);
    timer_destroy (timer);
    ufo_decoder_free(decoder);

    return error == EIO ? 0 : error;
}

int main(int argc, char const* argv[])
{
    int getopt_ret, index;

    enum {
        CLEAR_FRAME  = 'c',
        DRY_RUN      = 'd',
        FRAME_RATE   = 'f',
        HELP         = 'h',
        SET_NUM_ROWS = 'r', 
        VERBOSE      = 'v',
        CONTINUE,
        NUM_ROWS,
        CONVERT_BAYER,
    };

    static struct option long_options[] = {
        { "num-rows",           required_argument, 0, SET_NUM_ROWS },
        { "clear-frame",        no_argument, 0, CLEAR_FRAME },
        { "verbose",            no_argument, 0, VERBOSE },
        { "help",               no_argument, 0, HELP },
        { "dry-run",            no_argument, 0, DRY_RUN },
        { "print-frame-rate",   no_argument, 0, FRAME_RATE },
        { "continue",           no_argument, 0, CONTINUE },
        { "print-num-rows",     no_argument, 0, NUM_ROWS },
        { "convert-bayer",      no_argument, 0, CONVERT_BAYER },
        { 0, 0, 0, 0 }
    };

    static Options opts = {
        .rows = 1088,
        .verbose = 0,
        .dry_run = 0,
        .clear_frame = 0,
        .print_frame_rate = 0,
        .print_num_rows = 0,
        .cont = 0,
        .convert_bayer = 0
    };

    while ((getopt_ret = getopt_long(argc, (char *const *) argv, "r:cvhdf", long_options, &index)) != -1) {
        switch (getopt_ret) {
            case SET_NUM_ROWS:
                opts.rows = atoi(optarg);
                break;
            case CLEAR_FRAME:
                opts.clear_frame = 1;
                break;
            case VERBOSE:
                opts.verbose = 1;
                break;
            case HELP:
                usage();
                return 0;
            case DRY_RUN:
                opts.dry_run = 1;
                break;
            case FRAME_RATE:
                opts.print_frame_rate = 1;
                break;
            case CONTINUE:
                opts.cont = 1;
                break;
            case NUM_ROWS:
                opts.print_num_rows = 1;
                break;
            case CONVERT_BAYER:
                opts.convert_bayer = 1;
                break;
            default:
                break;
        }
    }

    if (optind == argc) {
        printf("ipedec: no input files\n");
        return 1;
    }

    while (optind < argc) {
        int errcode = process_file(argv[optind++], &opts);

        if (errcode != 0)
            return errcode;
    }

    return 0;
}



src/libipe.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include "libipe.h"
#include "libipe-private.h"
#include "config.h"
#include <xmmintrin.h>

#define IPECAMERA_NUM_CHANNELS 16
#define IPECAMERA_PIXELS_PER_CHANNEL 128
#define IPECAMERA_WIDTH (IPECAMERA_NUM_CHANNELS * IPECAMERA_PIXELS_PER_CHANNEL)


#define CHECK_VALUE(value, expected) \
    if (value != expected) { \
        fprintf(stderr, "<%s:%i> 0x%x != 0x%x\n", __FILE__, __LINE__, value, expected); \
        err = 1; \
    }

#define CHECK_FLAG(flag, check, ...) \
    if (!(check)) { \
        fprintf(stderr, "<%s:%i> Unexpected value 0x%x of " flag "\n", __FILE__, __LINE__,  __VA_ARGS__); \
        err = 1; \
    }


/**
 * \brief Setup a new decoder instance
 *
 * \param height Number of rows that are expected in the data stream
 * \param raw The data stream from the camera or NULL if set later with
 * ipe_decoder_set_raw_data.
 * \param num_bytes Size of the data stream buffer in bytes
 *
 * \return A new decoder instance that can be used to iterate over the frames
 * using ipe_decoder_get_next_frame.
 */
ipe_decoder ipe_decoder_new(uint32_t height, uint32_t *raw, size_t num_bytes)
{
    ipe_decoder decoder = malloc(sizeof(struct ipe_decoder_t));
    if (decoder == NULL)
        return NULL;

    decoder->height = height;
    ipe_decoder_set_raw_data(decoder, raw, num_bytes);
    return decoder;
}


/**
 * \brief Release decoder instance
 *
 * \param decoder An ipe_decoder instance
 */
void ipe_decoder_free(ipe_decoder decoder)
{
    free(decoder);
}


/**
 * \brief Set raw data stream
 *
 * \param decoder An ipe_decoder instance
 * \param raw Raw data stream
 * \param num_bytes Size of data stream buffer in bytes
 */
void ipe_decoder_set_raw_data(ipe_decoder decoder, uint32_t *raw, size_t num_bytes)
{
    decoder->raw = raw;
    decoder->num_bytes = num_bytes;
    decoder->current_pos = 0;
}

static int ipe_decode_frame(uint16_t *pixel_buffer, uint32_t *raw, int num_rows, int *offset)
{
    static int channel_order[IPECAMERA_NUM_CHANNELS] = { 15, 13, 14, 12, 10, 8, 11, 7, 9, 6, 5, 2, 4, 3, 0, 1 };

    int info;
    int row = 0;
    int channel = 0;
    int pos = 0;
    uint32_t data;
    const int bytes = 43;

#ifdef HAVE_SSE
    const uint32_t mask = 0x3FF;
    __m128i mmask = _mm_set_epi32(mask, mask, mask, mask);
    __m128i packed;
    __m128i tmp1, tmp2;
    uint32_t result[4] __attribute__ ((aligned (16))) = {0};
#endif

    do {
        info = raw[0];
        row = (info >> 4) & 0x7FF;
        int pixels = (info >> 20) & 0xFF;

        channel = channel_order[info & 0x0F];
        int base = row * IPECAMERA_WIDTH + channel * IPECAMERA_PIXELS_PER_CHANNEL;

#ifdef DEBUG
        int err = 0;
        int header = (info >> 30) & 0x03;   // 2 bits
        const int bpp = (info >> 16) & 0x0F;      // 4 bits
        CHECK_FLAG("raw header magick", header == 2, header);
        CHECK_FLAG("pixel size, only 10 bits are supported", bpp == 10, bpp);
        CHECK_FLAG("channel, limited by %i output channels", channel < IPECAMERA_NUM_CHANNELS, channel, IPECAMERA_NUM_CHANNELS);
#endif

        /* "Correct" missing pixel */
        if ((row < 2) && (pixels == (IPECAMERA_PIXELS_PER_CHANNEL - 1))) {
            pixel_buffer[base] = 0;
            /* base++; */
        }
#ifdef DEBUG
        else 
            CHECK_FLAG("number of pixels, %i is expected", pixels == IPECAMERA_PIXELS_PER_CHANNEL, pixels, IPECAMERA_PIXELS_PER_CHANNEL);
#endif

#ifdef HAVE_SSE
        for (int i = 1 ; i < bytes-4; i += 4, base += 12) {
            packed = _mm_set_epi32(raw[i], raw[i+1], raw[i+2], raw[i+3]);

            tmp1 = _mm_srli_epi32(packed, 20);
            tmp2 = _mm_and_si128(tmp1, mmask);
            _mm_storeu_si128((__m128i*) result, tmp2);
            pixel_buffer[base] = result[0];
            pixel_buffer[base+3] = result[1];
            pixel_buffer[base+6] = result[2];
            pixel_buffer[base+9] = result[3];

            tmp1 = _mm_srli_epi32(packed, 10);
            tmp2 = _mm_and_si128(tmp1, mmask);
            _mm_storeu_si128((__m128i*) result, tmp2);
            pixel_buffer[base+1] = result[0];
            pixel_buffer[base+4] = result[1];
            pixel_buffer[base+7] = result[2];
            pixel_buffer[base+10] = result[3];

            tmp1 = _mm_and_si128(packed, mmask);
            _mm_storeu_si128((__m128i*) result, tmp1);
            pixel_buffer[base+2] = result[0];
            pixel_buffer[base+5] = result[1];
            pixel_buffer[base+8] = result[2];
            pixel_buffer[base+11] = result[3];
        }
        
        /* Compute last pixels the usual way */
        for (int i = bytes-4; i < bytes; i++) {
            data = raw[i];
            pixel_buffer[base++] = (data >> 20) & 0x3FF;
            pixel_buffer[base++] = (data >> 10) & 0x3FF;
            pixel_buffer[base++] = data & 0x3FF;
        }
#else
        for (int i = 1 ; i < bytes; i++) {
            data = raw[i];
#ifdef DEBUG
            header = (data >> 30) & 0x03;   
            CHECK_FLAG("raw data magick", header == 3, header);
            if (err) 
                return err;
#endif
            pixel_buffer[base++] = (data >> 20) & 0x3FF;
            pixel_buffer[base++] = (data >> 10) & 0x3FF;
            pixel_buffer[base++] = data & 0x3FF;
        }
#endif

        data = raw[bytes];
#ifdef DEBUG
        header = (data >> 30) & 0x03;
        CHECK_FLAG("raw data magick", header == 3, header);
        CHECK_FLAG("raw footer magick", (data & 0x3FF) == 0x55, (data & 0x3FF));
        if (err) 
            return err;
#endif

        int ppw = pixels >> 6;
        for (int j = 0; j < ppw; j++)
            pixel_buffer[base++] = (data >> (10 * (ppw - j))) & 0x3FF;

        pos += bytes + 1;
        raw += bytes + 1;
    } while ((row < (num_rows - 1)) || (channel != 1));

    *offset = pos;
    return 0;
}


/**
 * \brief Deinterlace by interpolating between two rows
 *
 * \param in Input frame
 * \param out Destination of interpolated frame
 * \param width Width of frame in pixels
 * \param heigh Height of frame in pixels
 */
void ipe_deinterlace_interpolate(const uint16_t *in, uint16_t *out, int width, int height)
{
    const size_t row_size_bytes = width * sizeof(uint16_t);

    for (int row = 0; row < height; row++) {
        /* Copy one line */
        memcpy(out, in + row*width, row_size_bytes);
        out += width;

        /* Interpolate between source row and row+1 */ 
        for (int x = 0; x < width; x++) {
            out[x] = (int) (0.5 * in[row*width + x] + 0.5 * in[(row+1)*width + x]);
        }
        out += width;
    }

    /* Copy last row */
    memcpy(out, in + width * (height - 1), row_size_bytes);
}


/**
 * \brief Deinterlace by "weaving" the rows of two frames
 *
 * \param in1 First frame
 * \param in2 Second frame
 * \param out Destination of weaved frame
 * \param width Width of frame in pixels
 * \param heigh Height of frame in pixels
 */
void ipe_deinterlace_weave(const uint16_t *in1, const uint16_t *in2, uint16_t *out, int width, int height)
{
    const size_t row_size_bytes = width * sizeof(uint16_t);
    for (int row = 0; row < height; row++) {
        memcpy(out, in1 + row*width, row_size_bytes); 
        out += width;
        memcpy(out, in2 + row*width, row_size_bytes); 
        out += width;
    }
}


/**
 * \brief Iterate and decode next frame
 *
 * This function tries to decode the next frame in the currently set raw data
 * stream. 
 *
 * \param decoder An ipe_decoder instance
 * \param pixels If pointer with NULL content is passed, a new buffer is
 * allocated otherwise, this user-supplied buffer is used.
 * \param frame_number Frame number as reported in the header
 * \param time_stamp Time stamp of the frame as reported in the header
 *
 * \return 0 in case of no error, ENOSR if end of stream was reached, ENOMEM if
 * NULL was passed but no memory could be allocated, EILSEQ if data stream is
 * corrupt and EFAULT if pixels is a NULL-pointer.
 */
int ipe_decoder_get_next_frame(ipe_decoder decoder, uint16_t **pixels, uint32_t *frame_number, uint32_t *time_stamp)
{

    uint32_t *raw = decoder->raw;
    int err = 0;
    int pos = decoder->current_pos;
    int advance;
    const int num_words = decoder->num_bytes / 4;

    if (pixels == NULL)
        return EFAULT;

    if (pos >= num_words)
        return ENOSR; 

    if (num_words < 16)
        return EILSEQ;

    if (*pixels == NULL) {
        *pixels = (uint16_t *) malloc(IPECAMERA_WIDTH * decoder->height * sizeof(uint16_t));
        if (*pixels == NULL)
            return ENOMEM;
    }

#ifdef DEBUG
    CHECK_VALUE(raw[pos++], 0x51111111);
    CHECK_VALUE(raw[pos++], 0x52222222);
    CHECK_VALUE(raw[pos++], 0x53333333);
    CHECK_VALUE(raw[pos++], 0x54444444);
    CHECK_VALUE(raw[pos++], 0x55555555);
    CHECK_VALUE(raw[pos++], 0x56666666);
    CHECK_VALUE(raw[pos] >> 28, 0x5);
    *frame_number = raw[pos++] & 0xF0000000;
    CHECK_VALUE(raw[pos] >> 28, 0x5);
    *time_stamp = raw[pos++] & 0xF0000000;
    if (err)
        return EILSEQ;
#else
    *frame_number = raw[pos + 6] & 0xF0000000;
    *time_stamp = raw[pos + 7] & 0xF0000000;
    pos += 8;
#endif

    err = ipe_decode_frame(*pixels, raw + pos, decoder->height, &advance);
    if (err)
        return EILSEQ;

    pos += advance;

#ifdef DEBUG
    CHECK_VALUE(raw[pos++], 0x0AAAAAAA);
    CHECK_VALUE(raw[pos++], 0x0BBBBBBB);
    CHECK_VALUE(raw[pos++], 0x0CCCCCCC);
    CHECK_VALUE(raw[pos++], 0x0DDDDDDD);
    CHECK_VALUE(raw[pos++], 0x0EEEEEEE);
    CHECK_VALUE(raw[pos++], 0x0FFFFFFF);
    CHECK_VALUE(raw[pos++], 0x00000000);
    CHECK_VALUE(raw[pos++], 0x01111111);
#else
    pos += 8;
#endif

    /* if bytes left and we see fill bytes, skip them */
    if ((raw[pos] == 0x0) && (raw[pos+1] == 0x1111111)) {
        pos += 2;
        while ((pos < num_words) && ((raw[pos] == 0x89abcdef) || (raw[pos] == 0x1234567)))
            pos++;
    }

    decoder->current_pos = pos;
    return 0;
}



ipedec.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <assert.h>

#define IPECAMERA_MAX_LINES 1088
#define IPECAMERA_MAX_CHANNELS 16
#define IPECAMERA_PIXELS_PER_CHANNEL 128
#define IPECAMERA_WIDTH (IPECAMERA_MAX_CHANNELS * IPECAMERA_PIXELS_PER_CHANNEL)
#define IPECAMERA_HEIGHT 1088

int ipecamera_channel_order[IPECAMERA_MAX_CHANNELS] = { 15, 13, 14, 12, 10, 8, 11, 7, 9, 6, 5, 2, 4, 3, 0, 1 };

#define CHECK_VALUE(value, expected) \
    if (value != expected) { \
        fprintf(stderr, "<%s:%i> 0x%x != 0x%x\n", __FILE__, __LINE__, value, expected); \
        err = 1; \
    }

#define CHECK_FLAG(flag, check, ...) \
    if (!(check)) { \
        fprintf(stderr, "<%s:%i> Unexpected value 0x%x of " flag "\n", __FILE__, __LINE__,  __VA_ARGS__); \
        err = 1; \
    }


static int ipecamera_get_payload(uint16_t *pixel_buffer, 
        int line_req, size_t size, 
        uint32_t *payload,
        int *advance, int *increase_line)
{
    int ppw;
    int err = 0;

    uint32_t info = payload[0];
    int channel = info & 0x0F;          // 4 bits
    int line = (info >> 4) & 0x7FF;     // 11 bits
    int bpp = (info >> 16) & 0x0F;      // 4 bits
    int pixels = (info >> 20) & 0xFF;   // 8 bits
    int header = (info >> 30) & 0x03;   // 2 bits

    int bytes, pix, pix_offset = 0;
    uint32_t data;

    *increase_line = channel == 15 ? 1 : 0;
    channel = ipecamera_channel_order[channel];

    CHECK_FLAG("payload header magick", header == 2, header);
    CHECK_FLAG("pixel size, only 10 bits are supported", bpp == 10, bpp);
    CHECK_FLAG("row number, should be %i", line == line_req, line, line_req);
    CHECK_FLAG("channel, limited by %i output channels", channel < IPECAMERA_MAX_CHANNELS, channel, IPECAMERA_MAX_CHANNELS);

    if ((line < 2) && (pixels == (IPECAMERA_PIXELS_PER_CHANNEL - 1))) {
        pix_offset = 1;
        pixel_buffer[channel*IPECAMERA_PIXELS_PER_CHANNEL] = 0;
    } 
    else 
        CHECK_FLAG("number of pixels, %i is expected", pixels == IPECAMERA_PIXELS_PER_CHANNEL, pixels, IPECAMERA_PIXELS_PER_CHANNEL);

    bytes = pixels / 3;
    ppw = pixels - bytes * 3;
    if (ppw) 
        ++bytes;

    CHECK_FLAG("payload data bytes, at least %i are expected", bytes < size, (unsigned int) size, bytes);

    if (err) 
        return err;

    pix = pix_offset;
    for (int i = 1; i < bytes; i++) {
        data = payload[i];
        header = (data >> 30) & 0x03;   

        CHECK_FLAG("payload data magick", header == 3, header);
        if (err) 
            return err;

        pixel_buffer[channel*IPECAMERA_PIXELS_PER_CHANNEL + pix++] = (data >> 20) & 0x3FF;
        pixel_buffer[channel*IPECAMERA_PIXELS_PER_CHANNEL + pix++] = (data >> 10) & 0x3FF;
        pixel_buffer[channel*IPECAMERA_PIXELS_PER_CHANNEL + pix++] = data & 0x3FF;
    }

    data = payload[bytes];
    header = (data >> 30) & 0x03;

    CHECK_FLAG("payload data magick", header == 3, header);
    CHECK_FLAG("payload footer magick", (data&0x3FF) == 0x55, (data&0x3FF));
    if ((err) && ((data & 0x3FF) != 0x55)) {
        err = 0; 
        int counter = 0;
        printf(" -> %x\n", data);
        while ((data & 0x3FF) != 0x55) {
            bytes++; 
            data = payload[bytes];
            counter++;
        }
        printf("skipped %i bytes: found %x\n", counter, data);
    }
    if (err) 
        return err;

    ppw = pixels % 3;
    assert(ppw < 3);

    for (int j = 0; j < ppw; j++, pix++)
        pixel_buffer[channel*IPECAMERA_PIXELS_PER_CHANNEL + pix] = (data >> (10 * (ppw - j))) & 0x3FF;

    *advance = bytes + 1;
    return 0;
}

static int ipecamera_parse_image(int first_line, int num_lines, 
        size_t size, uint32_t *linebuf, int *frames_decoded) 
{
    int err = 0, current_frame = 0;
    int pos = 0, advance, increase_line;

    if (size < 16) {
        fprintf(stderr, "The payload is tool small, we should have at least 8 header dwords and 8 footer.");
        return 1;
    }

    FILE *fp = fopen("image.dec.raw", "wb");
    uint16_t *pixel_buffer = (uint16_t *) malloc(num_lines * IPECAMERA_WIDTH * sizeof(uint16_t));

    printf("Decoding");
    while (size > 0) {
        printf(".");
        int line = first_line;
        CHECK_VALUE(linebuf[pos++], 0x51111111);
        CHECK_VALUE(linebuf[pos++], 0x52222222);
        CHECK_VALUE(linebuf[pos++], 0x53333333);
        CHECK_VALUE(linebuf[pos++], 0x54444444);
        CHECK_VALUE(linebuf[pos++], 0x55555555);
        CHECK_VALUE(linebuf[pos++], 0x56666666);
        CHECK_VALUE(linebuf[pos++], 0x57777777);
        CHECK_VALUE(linebuf[pos++], 0x58888888);
        if (err) 
            return err;

        size -= 16;

        while ((size > 0) && (line < (first_line + num_lines))) {
            err = ipecamera_get_payload(pixel_buffer + line * IPECAMERA_WIDTH,
                    line - first_line, size, linebuf + pos, &advance, &increase_line);
            if (err) 
                return err;

            pos += advance;
            size -= advance;
            line += increase_line;
        }

        CHECK_FLAG("lines read, we expect to read exactly %i lines", line == (first_line + num_lines), line - first_line, num_lines);

        CHECK_VALUE(linebuf[pos++], 0x0AAAAAAA);
        CHECK_VALUE(linebuf[pos++], 0x0BBBBBBB);
        CHECK_VALUE(linebuf[pos++], 0x0CCCCCCC);
        CHECK_VALUE(linebuf[pos++], 0x0DDDDDDD);
        CHECK_VALUE(linebuf[pos++], 0x0EEEEEEE);
        CHECK_VALUE(linebuf[pos++], 0x0FFFFFFF);
        CHECK_VALUE(linebuf[pos++], 0x00000000);
        CHECK_VALUE(linebuf[pos++], 0x01111111);

        fwrite(pixel_buffer, IPECAMERA_WIDTH * num_lines, sizeof(uint16_t), fp);

        if (linebuf[pos] == 0)
            break;

        current_frame++;
    }
    fclose(fp);
    free(pixel_buffer);
    printf("\n");
    *frames_decoded = current_frame + 1;
    return err;
}

int main(int argc, char const* argv[])
{
    if (argc < 3) {
        fprintf(stderr, "Usage: ipedec <filename> <number of lines per frame>\n");
        return 1;
    }

    FILE *fp = fopen(argv[1], "rb"); 
    if (fp == NULL) {
        fprintf(stderr, "Couldn't open file %s", argv[1]);
        return 1;
    }

    const int lines_per_frame = atoi(argv[2]);

    fseek(fp, 0, SEEK_END);
    const size_t length = ftell(fp);
    rewind(fp);

    char *buffer = (char *) malloc(length);
    if (buffer == NULL) {
        fclose(fp);
        return 1;
    }

    size_t buffer_length = fread(buffer, 1, length, fp);
    fclose(fp);
    if (buffer_length != length) {
        free(buffer);
        return 1;
    }

    int num_frames;
    if (ipecamera_parse_image(0, lines_per_frame, length, (uint32_t *) buffer, &num_frames))
        fprintf(stderr, "Couldn't parse image\n");

    printf("Decoded %i frames.\n", num_frames);
    free(buffer);
    return 0;
}

